### ISSUE TITLE
ICE: `Type system constant with non valtree'able type evaluated but no error emitted`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#133028]
auto-reduced (treereduce-rust):
````rust
#![feature(associated_const_equality)]

trait Trait {
    const F: fn();
}

fn take(_: impl Trait<F = { || {} }>) {}

fn main() {}

````

original:
````rust
#![feature(associated_const_equality)]

trait Trait { const F: fn(); }

fn take(_: impl Trait<F = { || {} }>) {
    if let RCZ = &raw const *&0 { }
}

fn main() {}

````

Version information
````
rustc 1.84.0-nightly (22bcb81c6 2024-11-14)
binary: rustc
commit-hash: 22bcb81c6636f8ccc24478f92b30bf9266f116ef
commit-date: 2024-11-14
host: x86_64-unknown-linux-gnu
release: 1.84.0-nightly
LLVM version: 19.1.3

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
warning: trait `Trait` is never used
 --> /tmp/icemaker_global_tempdir.mt4XPzB9Nw7Z/rustc_testrunner_tmpdir_reporting.4x5gPz1Id0gX/mvce.rs:3:7
  |
3 | trait Trait {
  |       ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: function `take` is never used
 --> /tmp/icemaker_global_tempdir.mt4XPzB9Nw7Z/rustc_testrunner_tmpdir_reporting.4x5gPz1Id0gX/mvce.rs:7:4
  |
7 | fn take(_: impl Trait<F = { || {} }>) {}
  |    ^^^^

warning: 2 warnings emitted

note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error: Type system constant with non valtree'able type evaluated but no error emitted
  |
  = note: delayed at compiler/rustc_trait_selection/src/traits/mod.rs:613:39 - disabled backtrace

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.84.0-nightly (22bcb81c6 2024-11-14) running on x86_64-unknown-linux-gnu

query stack during panic:
end of query stack

```

</p>
</details>

@rustbot label +F-associated_const_equality

**Comment by fmease** [comment]
Literally #119783
